thor/main.py

```diff
@@ -1,193 +1,520 @@
+        )
+
+    return CheckpointData(
+        stage="range_and_transform", filtered_observations=filtered_observations
+    )
+
+
+@dataclass
+class LinkTestOrbitStageResult:
+    """
+    Result of a single stage of the THOR pipeline.
+    """
+
+    name: Literal[
+        "filter_observations",
+        "range_and_transform",
+        "cluster_and_link",
+        "initial_orbit_determination",
+        "differential_correction",
+        "recover_orbits",
+    ]
+    result: Iterable[Any]
+    path: Optional[Iterable[str]] = None
+
+
 def link_test_orbit(
     test_orbit: TestOrbit,
     observations: Observations,
+    working_dir: Optional[str] = None,
     filters: Optional[List[ObservationFilter]] = None,
     config: Optional[Config] = None,
-) -> Iterator[Any]:
+    allow_config_override: bool = False,
+) -> Iterator[LinkTestOrbitStageResult]:
     """
     Run THOR for a single test orbit on the given observations. This function will yield
     results at each stage of the pipeline.
-        1. transformed_detections
-        2. clusters, cluster_members
-        3. iod_orbits, iod_orbit_members
-        4. od_orbits, od_orbit_members
-        5. recovered_orbits, recovered_orbit_members
+        1. filtered observations
+        2. transformed_detections
+        3. clusters, cluster_members
+        4. iod_orbits, iod_orbit_members
+        5. od_orbits, od_orbit_members
+        6. recovered_orbits, recovered_orbit_members
 
     Parameters
     ----------
     test_orbit : `~thor.orbit.TestOrbit`
         Test orbit to use to gather and transform observations.
     observations : `~thor.observations.observations.Observations`
         Observations from which range and transform the detections.
+    working_dir : str, optional
+        Directory with persisted config and checkpointed results.
     filters : list of `~thor.observations.filters.ObservationFilter`, optional
         List of filters to apply to the observations before running THOR.
-    propagator : `~adam_core.propagator.propagator.Propagator`
-        Propagator to use to propagate the test orbit and generate
-        ephemerides.
-    max_processes : int, optional
-        Maximum number of processes to use for parallelization.
+    config : `~thor.config.Config`, optional
+        Configuration to use for THOR. If None, the default configuration will be used.
+    allow_config_override : bool, optional
+        If using a checkpoint directory, setting this to True will alllow the configuration
+        to not match the previous configuration.
     """
     time_start = time.perf_counter()
     logger.info("Running test orbit...")
 
+    test_orbit_directory = None
+    if working_dir is not None:
+        working_dir = pathlib.Path(working_dir)
+        logger.info(f"Using working directory: {working_dir}")
+        test_orbit_directory = pathlib.Path(working_dir, test_orbit.orbit_id)
+        test_orbit_directory.mkdir(parents=True, exist_ok=True)
```

Comment:
One small bug: this doesn't appear to create the "inputs" folder within working_dir so checking if the config exists in the subsequent function call fails when it tries to write a new config. 

Could add something like:
```python
inputs_dir = pathlib.Path(working_dir, "inputs")
inputs_dir.mkdir(parents=True, exist_ok=True)
```